domain/storageprovisioning/state/filesystem.go

```diff
@@ -0,0 +1,288 @@
+	JOIN   storage_filesystem sf ON sfa.storage_filesystem_uuid = sf.uuid
+	JOIN   machine m ON sfa.net_node_uuid = m.net_node_uuid
+	WHERE  sfa.uuid IN ($filesystemAttachmentUUIDs[:])
+	UNION
+	SELECT      sfa.uuid,
+	            sf.filesystem_id,
+		        NULL AS machine_name,
+		        u.name AS unit_name
+	FROM        storage_filesystem_attachment sfa
+	JOIN        storage_filesystem sf ON sfa.storage_filesystem_uuid = sf.uuid
+	LEFT JOIN   machine m ON sfa.net_node_uuid != m.net_node_uuid
+	JOIN        unit u ON sfa.net_node_uuid = u.net_node_uuid
+	WHERE       sfa.uuid IN ($filesystemAttachmentUUIDs[:])
+)
+`
+
+	uuidToIDsStmt, err := st.Prepare(q, filesystemAttachmentIDs{}, uuidInputs)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var dbVals []filesystemAttachmentIDs
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, uuidToIDsStmt, uuidInputs).GetAll(&dbVals)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		} else if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	rval := make(map[string]storageprovisioning.FilesystemAttachmentID, len(dbVals))
+	for _, v := range dbVals {
+		id := storageprovisioning.FilesystemAttachmentID{
+			FilesystemID: v.FilesystemID,
+		}
+		if v.MachineName.Valid {
+			id.MachineName = &v.MachineName.String
+		}
+		if v.UnitName.Valid {
+			id.UnitName = &v.UnitName.String
+		}
+
+		rval[v.UUID] = id
+	}
+	return rval, nil
+}
+
+// GetFilesystemAttachmentLifeForNetNode returns a mapping of filesystem
+// attachment uuids to the current life value for each machine provisioned
+// filesystem attachment that is to be provisioned by the machine owning the
+// supplied net node.
+func (st *State) GetFilesystemAttachmentLifeForNetNode(
+	ctx context.Context,
+	netNodeUUID string,
+) (map[string]life.Life, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	return st.getFilesystemAttachmentLifeForNetNode(ctx, db, netNodeUUID)
+}
+
+// getFilesystemAttachmentLifeForNetNode returns a mapping of filesystem
+// attachment uuids to the current life value for each machine provisioned
+// filesystem attachment that is to be provisioned by the machine owning the
+// supplied net node.
+func (st *State) getFilesystemAttachmentLifeForNetNode(
+	ctx context.Context,
+	db domain.TxnRunner,
+	netNodeUUID string,
+) (map[string]life.Life, error) {
+	netNodeInput := netNodeUUIDRef{UUID: netNodeUUID}
+	stmt, err := st.Prepare(`
+SELECT DISTINCT &attachmentLife.*
+FROM            storage_filesystem_attachment
+WHERE           provision_scope_id=1
+AND             net_node_uuid=$netNodeUUIDRef.net_node_uuid
+		`, attachmentLife{}, netNodeInput)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	var fsAttachmentLives attachmentLives
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		exists, err := st.checkNetNodeExists(ctx, tx, netNodeUUID)
```

Comment:
Do we need to do the existence check here?
If the caller does want to know if the net node exists or not, we should return a NetNodeNotFound then.

Reply:
This is done inside of watcher. No caller will ever deal with this error type nor do we have a case today where this is needed. The check is done to fail the call incase it is orphaned somehow. i.e we are asserting our pre-conditions are correct. Remember that transactions are reentrant so we can not rely on the previously held assumptions.